### PR TITLE
refactor speed up/down range

### PR DESCRIPTION
**Pull request by Rechi**
## Description
move the hardcoded values for minimal speed and maximal speed into functions

## Motivation and Context
The same two float values were hardcoded in 11 different lines. 

## How Has This Been Tested?
Runtime tested

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/GUIInfoManager.cpp [hunk 1]
```diff
@@ -7174,7 +7174,7 @@ bool CGUIInfoManager::GetBool(int condition1, int contextWindow, const CGUIListI
       bReturn = g_application.m_pPlayer->GetPlaySpeed() < 0;
       break;
     case PLAYER_FORWARDING:
-      bReturn = g_application.m_pPlayer->GetPlaySpeed() > 1.5;
+      bReturn = g_application.m_pPlayer->GetPlaySpeed() > IPlayer::MaxPlaySpeed();
```
Comment (FernetMenta) [inline xbmc/GUIInfoManager.cpp]: this is a change of behavior

File: xbmc/GUIInfoManager.cpp [hunk 2]
```diff
@@ -7954,7 +7954,7 @@ std::string CGUIInfoManager::GetMultiInfoLabel(const GUIInfo &info, int contextW
   {
     std::string strTime;
     float speed = g_application.m_pPlayer->GetPlaySpeed();
-    if (speed < 0.8 || speed > 1.5)
```
Comment (FernetMenta) [inline xbmc/GUIInfoManager.cpp]: changes behavior

File: xbmc/GUIInfoManager.cpp (hunk 1 above)
Comment (Rechi) [inline xbmc/GUIInfoManager.cpp]: should I replace it with `IPlayer::MaxPlaySpeed() + 0.05`?

Comment (FernetMenta) [inline xbmc/GUIInfoManager.cpp]: better change a couple of things:

- this feature is referred to as tempo, not playspeed
- min / max are currently 0.8 / 1.5
- the diff of .5 is there in order to deal with rounding errors, would be nice if the code reflects this

**Comment by Rechi** [comment]
updated

File: xbmc/GUIInfoManager.cpp (hunk 1 above)
Comment (Rechi) [inline xbmc/GUIInfoManager.cpp]: I know setting this to `1.55` would changed behavior, but I don't understand why rounding errors must not be considered here.

Comment (FernetMenta) [inline xbmc/GUIInfoManager.cpp]: > but I don't understand why rounding errors must not be considered here.

me neither. I think they should. so it changes behaviour into the right direction :)

Comment (Rechi) [inline xbmc/GUIInfoManager.cpp]: actually I've update `IPlayer::MaxPlaySpeed()` to return `1.5`, so should I add `+ 0.05` like I've done at the other places?

File: xbmc/GUIInfoManager.cpp (hunk 2 above)
Comment (Rechi) [inline xbmc/GUIInfoManager.cpp]: should rounding errors also be considered here?

File: xbmc/GUIInfoManager.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/GUIInfoManager.cpp]: sounds reasonable. the difference between 1.5x and ff is that on ff the indicators stay on screen.

File: xbmc/GUIInfoManager.cpp (hunk 2 above)
Comment (FernetMenta) [inline xbmc/GUIInfoManager.cpp]: yes, I think so

**Comment by FernetMenta** [comment]
cleanup brings some issues to surface. I think you can add the label "fix" to this pr. well done

**Comment by Paxxi** [comment]
Mark the methods constexpr as well, compiler should basically treat them as defines in that case

**Comment by Rechi** [comment]
must there still be something marked with `const` when marking the function as `constexpr`?

**Comment by da-anda** [comment]
hmm, instead of hardcoded values in 11 places we now have the same logic duplicated in 11 different places (`IPlayer::MinTempo() - 0.05`). Sorry for the probably dump question, but if every usecase of `IPlayer::MinTempo()` always will get 0.05 substracted, why not simply change the initial value of it? Same ofc with MaxTempo().
Also, what about having a function like "IPlayer::isWithinPlaySpeedBoundaries(playSpeed)" instead of having to duplicate this `playSpeed >= IPlayer::MinTempo() - 0.05 && playSpeed <= IPlayer::MaxTempo() + 0.05` logic in dozens of places?

**Comment by FernetMenta** [comment]
> Also, what about having a function like "IPlayer::isWithinPlaySpeedBoundaries(playSpeed)"

not a bad idea, but

- the correct term is tempo not playspeed
- should not be a static method of IPlayer, paplayer only supports tempo 1, external player does not support tempo either

**Comment by Rechi** [comment]
have update with 3 new functions
- IsPlayingTempo
- MinTempo
- MaxTempo

Now I've an issue with getting the Min/MaxTempo into this line https://github.com/xbmc/xbmc/blob/958d453/xbmc/cores/VideoPlayer/VideoPlayerAudio.cpp#L41. Maybe anyone has an idea.

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 3]
```diff
@@ -3672,6 +3677,16 @@ bool CVideoPlayer::SupportsTempo()
   return m_canTempo;
 }
 
+const float CVideoPlayer::MinTempo() const
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp]: this is not const, actually this can change during playback

**Comment by FernetMenta** [comment]
Now this gets beyond a simple refactoring so we can change more things to get this right. VP moves away from the old polling model. That is that GUI bothers player with many calls and locks every cycle to get data that have not changed. playspeed does even lack proper locking.

Add playspeed to CProcessInfo and push updates also to CDataCacheCore. GUI can poll CDataCacheCore instead of player. VIdeoPlayerAudio has a ref to ProcessInfo so it gets access to required info.

**Comment by Rechi** [comment]
@FernetMenta is this superseded by #12470?

**Comment by FernetMenta** [comment]
yes, it is

**Pull request closed without merging**